src/sdk/namespace_fs.js

```diff
@@ -513,7 +534,36 @@ class NamespaceFS {
                 if (r.common_prefix) {
                     res.common_prefixes.push(r.key);
                 } else {
-                    res.objects.push(this._get_object_info(bucket, r.key, r.stat, 'null'));
+                    const obj_info = this._get_object_info(bucket, r.key, r.stat, 'null', true);
+                    if (!list_versions && obj_info.delete_marker) {
+                        continue;
+                    }
+                    if (list_versions) {
+                        const obj_info_list = [];
+                        versions.forEach(v => {
+                            let search_key = v.key.replace(HIDDEN_VERSIONS_PATH, '');
```

Comment:
if I understood correctly -  key is not the path anymore,  can this code be removed?

Reply:
so on get_entry_key() dir_key is dir1/.versions/ and ent is key1_xyz? and then we remove .versions/, do you think we can just return dir1/key1_xyz directly from get_entry_key() instead ?

Reply:
yes but you will need to remove it only from the end of dir_key so you won't have these many cases, no? and also I think it's the correct place to put it since this is the function handling how get_entry_key() should look like

Reply:
One inevitable case. we need `full_path` of the `version file` to get the stat of the `versioned object`. If we modify the `r.key` in `get_entry_key()` then for `stat` will result in taking `stat` on latest object.

Please correct me if wrong,